docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,152 @@
+images between different stages (eg. from "development" to "production").
+This proposal also describes priciples a way how to configure OpenShift to serve as
+a Continuous Delivery tool.
+
+## Constraints and Assumptions
+
+This document describes the following ways to promote images
+
+* Based on human intervention
+* Driven purely by OpenShift
+* Based on external Continuous Delivery tool (Jenkins, go.cd, Snap, ...)
+
+Docker images can be promoted (and the use-cases are valid for):
+
+* Within a single Project
+* Between multiple Projects
+* Between multiple Clusters
+
+The Docker image promotion can be done using the Docker image tags and LABELs.
+You can also use other metadata that OpenShift adds to the Image resource,
+such as annotations or labels, to allow conditional promotion, but this document
+does not describe these advanced scenarios.
+
+Another constraint is that in some cases the image promotion might not be enough,
+and you want to promote a set of images or entire deployment as one single unit
+of promotion.
+
+## Use Cases
+
+### Automatic promotion based on running tests in Jenkins
+
+As a developer of a Rails application, when I update the source code and push it
+to a remote source code repository, I want Jenkins to detect this change and perform a S2I
+build of my application in OpenShift. Depending of the build result, I want Jenkins to
+promote the newly built Docker image by tagging it as "qa-ready" and use OpenShift
+deployments to promote it as "production-ready".
+
+#### Example
+
+As a user I create two DeploymentConfig's:
+
+* "frontend-test"
+* "frontend-production"
+
+The "frontend-test" DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the "qa-ready" ImageStreamTag. This DeploymentConfig also defines the "post" lifecycle
+hook that executes a script in the container deployed and performs additional verification
+and tests.
+
+Example of DeploymentConfig post-lifecycle hook:
+```json
+"strategy": {
+  "type": "Rolling",
+  "rollingParams": {
+    "pre": {
+      "failurePolicy": "Abort",
+      "execNewPod": {
+        "containerName": "test-container",
+        "command": [
+          "/opt/app-root/bin/verify-deployment"
+        ],
+      }
+    }
+  }
+}
+```
+
+Based on the result of the `verify-deployment` command, that runs in the application
+image, the command will promote the image by tagging it as `production-ready` (eg. using
+the `oc tag` command). The command might also notify me about the verification failure.
+
+The "frontend-production" DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the "prod-ready" ImageStreamTag. When a new Docker image with this tag is available,
+it can automatically rolls out a new production deployment. In case you don't want to roll out
+the "frontend-production" automatically, you can avoid specifying the ImageChangeTrigger and
+roll out the new deployment by hand.
```

Comment:
@bparees I thought that if you do `oc tag` it will create a new tag for the image. This example assume that you are using internal registry for this, so you don't necessarily need to push.

@bparees you're saying that creating a new tag in ImageStream and having ImageChangeTrigger in DC won't trigger a new deployment?